File: Excel/400-499/492/492 Challenge.py
import pandas as pd

path = "492 Date Shift Cipher Decrypter.xlsx"

input = pd.read_excel(path, usecols="A:B", nrows = 7)
test = pd.read_excel(path, usecols="C", nrows = 7)

def decrypt_date_cipher(text, date):
    repeat_date = (str(date) * ((len(text) // len(str(date))) + 1))[:len(text)]
    text_nums = [ord(char.lower()) - ord('a') for char in text]
    key_nums = [int(char) for char in repeat_date]
    decrypted_nums = [(text_num - key_num + 26) % 26 + ord('a') for text_num, key_num in zip(text_nums, key_nums)]
    decrypted_text = ''.join(chr(num) for num in decrypted_nums)
    return decrypted_text

input["Answer Expected"] = input.apply(lambda x: decrypt_date_cipher(x[0], x[1]), axis=1)

print(input["Answer Expected"].equals(test["Answer Expected"])) # True

File: Excel/400-499/492/492 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/492 Date Shift Cipher Decrypter.xlsx"

input = read_xlsx(path, range = "A1:B7")
test  = read_xlsx(path, range = "C1:C7")

decrypt_date_cipher <- function(text, date) {
  key <- str_sub(str_c(rep(date, ceiling(nchar(text) / nchar(date))), collapse = ""), 1, nchar(text))
  text_nums <- utf8ToInt(str_to_lower(text)) - utf8ToInt("a")
  key_nums <- key %>%
    str_split("") %>%
    flatten_chr() %>%
    as.integer()
  decrypted_nums <- map2_int(text_nums, key_nums, ~ (.x - .y + 26) %% 26 + utf8ToInt("a"))
  decrypted_text <- intToUtf8(decrypted_nums)
  return(decrypted_text)
}

output = input  %>%
  mutate(`Answer Expected` = map2_chr(Message, Key, decrypt_date_cipher))

identical(output$`Answer Expected`, test$`Answer Expected`)         
#> [1] TRUE 